services/user/InviteSys/plugin/src/lib.rs

```diff
@@ -0,0 +1,196 @@
+        Err(NotYetImplemented.err("set_blacklist"))
+    }
+}
+
+impl Invitee for Component {
+    fn accept(_id: InviteId) -> Result<(), CommonTypes::Error> {
+        Err(NotYetImplemented.err("accept_with_existing_account"))
+        // The thinking for only a single accept method is that the invite ID is passed
+        // to the account plugin. If the invite contains a valid invite private key
+        // then the login page can also show a Create Account button.
+    }
+
+    fn reject(_id: InviteId) -> Result<(), CommonTypes::Error> {
+        Err(NotYetImplemented.err("reject"))
+    }
+
+    fn decode_invite(id: InviteId) -> Result<Invite, CommonTypes::Error> {
+        let decoded = match URL_SAFE.decode(id.to_owned()) {
+            Ok(enc) => match String::from_utf8(enc) {
+                Ok(dec) => dec,
+                Err(_) => {
+                    return Err(DecodeInviteError.err("Error converting from UTF8"));
+                }
+            },
+            Err(_) => {
+                return Err(DecodeInviteError.err("Error decoding base64"));
+            }
+        };
+        let decoded: InviteParams = match serde_json::from_str(&decoded) {
+            Ok(d) => d,
+            Err(_) => {
+                return Err(DecodeInviteError.err("Error deserializing JSON string into object"));
+            }
+        };
+        let url = format!("{}/graphql", client::my_service_origin()?);
+        let pubkey = &decoded.pk;
+        let query = format!(
+            r#"query {{
+                getInvite(pubkey: "{pubkey}") {{
+                    pubkey,
+                    inviter
+                }}
+            }}"#,
+            pubkey = pubkey
+        );
+        let response: ResponseRoot = match server::post_graphql_get_json(&url, &query) {
+            Ok(o) => match serde_json::from_str(&o) {
+                Ok(parsed) => parsed,
+                Err(e) => {
+                    return Err(QueryError.err(&e.to_string()));
+                }
+            },
+            Err(e) => {
+                return Err(QueryError.err(&e.message));
+            }
+        };
+        let inv = match response.data.getInvite {
+            Some(inv) => inv,
+            None => {
+                return Err(QueryError.err("Invite not found"));
+            }
+        };
+
+        if inv.inviter != decoded.inviter {
+            return Err(CorruptedInviteId.err(&id));
+        }
+
+        Ok(Invite {
+            inviter: decoded.inviter,
+            app: decoded.app,
+            callback: decoded.cb,
+        })
+    }
+}
+
+impl Inviter for Component {
+    fn generate_invite(callback_subpath: String) -> Result<Url, CommonTypes::Error> {
+        let inviter = accounts::get_logged_in_user()?;
+
+        let inviter = match inviter {
+            Some(i) => i,
+            None => {
+                return Err(InviterLoggedIn.err(""));
+            }
+        };
```

Comment:
We could use [ok_or](https://doc.rust-lang.org/std/option/enum.Option.html#method.ok_or) here hey?

Reply:
Well... I was going to say no, because I'm returning `Err` from the function (not assigning it to inviter). 
Buuuut, actually the answer is yes. I can just use `?` ...clever :)

Actually, `ok_or_else` (`ok_or` is apparently eagerly evaluated?)

Reply:
Fixed (and in general made all the error handling more idiomatic)